nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/iot/AbstractAWSIoTProcessor.java

```diff
@@ -0,0 +1,230 @@
+        awsRegion = context.getProperty(REGION).getValue();
+        awsClientId = context.getProperty(PROP_CLIENT).isSet() ? context.getProperty(PROP_CLIENT).getValue() : PROP_DEFAULT_CLIENT;
+        awsKeepAliveSeconds = context.getProperty(PROP_KEEPALIVE).isSet() ? context.getProperty(PROP_KEEPALIVE).asInteger() : PROP_DEFAULT_KEEPALIVE;
+        awsTopic = context.getProperty(PROP_TOPIC).getValue();
+        awsQos = context.getProperty(PROP_QOS).isSet() ? context.getProperty(PROP_QOS).asInteger() : DEFAULT_QOS;
+        // initialize and connect to mqtt endpoint
+        mqttClient = connect(context);
+    }
+
+    @OnStopped
+    public void onStopped(final ProcessContext context) {
+        try {
+            mqttClient.disconnect();
+        } catch (MqttException me) {
+            getLogger().warn("MQTT " + me.getMessage());
+        }
+        getLogger().info("Disconnected");
+    }
+
+    /**
+     * Returns the lifetime-seconds of the established websocket-connection
+     * @return seconds
+     */
+    long getConnectionDuration() {
+        return dtLastConnect != null
+                ? TimeUnit.MILLISECONDS.toSeconds(new Date().getTime() - dtLastConnect.getTime())
+                : awsKeepAliveSeconds + 1;
+    }
+
+    /**
+     * In seconds get the remaining lifetime of the connection. It is not the actual time to
+     * expiration but an advice to when it is worth renewing the connection.
+     * @return seconds
+     */
+    long getRemainingConnectionLifetime() {
+        return awsKeepAliveSeconds - DEFAULT_CONNECTION_RENEWAL_BEFORE_KEEP_ALIVE_EXPIRATION;
+    }
+
+    /**
+     * Indicates if WebSocket connection is about to expire. It gives the caller an advice
+     * to renew the connection some time before the actual expiration.
+     * @return Indication (if true caller should renew the connection)
+     */
+    boolean isConnectionAboutToExpire() {
+        return getConnectionDuration() > getRemainingConnectionLifetime();
+    }
+
+    /**
+     * Connects to the websocket-endpoint over an MQTT client.
+     * @param context processcontext
+     * @return websocket connection client
+     */
+    MqttWebSocketAsyncClient connect(ProcessContext context) {
+        getCredentialsProvider(context).refresh();
+        AWSCredentials awsCredentials = getCredentialsProvider(context).getCredentials();
+        MqttWebSocketAsyncClient _mqttClient = null;
+
+        // generate mqtt endpoint-address with authentication details
+        String strEndpointAddress;
+        try {
+            strEndpointAddress = AWS4Signer.getAddress(awsRegion, awsEndpoint, awsCredentials);
+        } catch (Exception e) {
+            getLogger().error("Error while generating AWS endpoint-address caused by " + e.getMessage());
+            return null;
+        }
+        // extend clientId with random string in order to ensure unique id per connection
+        String clientId = awsClientId + RandomStringUtils.random(12, true, false);
```

Comment:
I'm on the fence about appending a random string to the client ID to ensure uniqueness. This could lead to laziness when creating the clientId and making it so the user doesn't know where a connection came from (when looking at the AWS UI). Wouldn't it be better to not append the random string and instead force a user to create a better clientIds?